packages/transport/src/types/typed-emitter.ts

```diff
@@ -0,0 +1,26 @@
+import { EventEmitter } from 'events';
+
+type EventMap = Record<string, any>;
+type EventKey<T extends EventMap> = string & keyof T;
+type EventReceiver<T> = (params: T) => void;
+
+interface Emitter<T extends EventMap> {
+    on<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>): void;
+    off<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>): void;
+    emit<K extends EventKey<T>>(eventName: K, params: T[K]): void;
+}
+
+export class TypedEmitter<T extends EventMap> implements Emitter<T> {
+    private emitter = new EventEmitter();
+    on<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>) {
+        this.emitter.on(eventName, fn);
+    }
+
+    off<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>) {
+        this.emitter.off(eventName, fn);
+    }
+
+    emit<K extends EventKey<T>>(eventName: K, params: T[K]) {
+        this.emitter.emit(eventName, params);
+    }
+}
```

Comment:
Is there a reason why the real emitter should be nested, or could it be simplified like this?
```suggestion
export interface TypedEmitter<T extends EventMap> {
    on<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>): this;
    off<K extends EventKey<T>>(eventName: K, fn: EventReceiver<T[K]>): this;
    emit<K extends EventKey<T>>(eventName: K, params: T[K]): boolean;
}

export class TypedEmitter<T extends EventMap> extends EventEmitter {}
```